SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -21,62 +21,113 @@ SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const PhysicalWorkerName&
+{
+	if (AuthOp.component_id != SpatialConstants::ENTITY_ACL_COMPONENT_ID)
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning, TEXT("Loadbalancer informed of authority change for entity %lld that was not related to the ACL component."), AuthOp.entity_id);
+		return;
+	}
+
+	if (AclAssignmentRequestIsQueued(AuthOp.entity_id) && AuthOp.authority == WORKER_AUTHORITY_NOT_AUTHORITATIVE)
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
+			TEXT("ACL authority lost for entity %lld. Can no longer enforce the previous request for this entity."),
+			AuthOp.entity_id);
+		return;
+	}
+	if (CanEnforce(AuthOp.entity_id))
+	{
+		MaybeQueueAclAssignmentRequest(AuthOp.entity_id);
 	}
 }
 
-// QueueAclAssignmentRequest is called from three places.
+// MaybeQueueAclAssignmentRequest is called from three places.
 // 1) AuthorityIntent change - Intent is not authoritative on this worker - ACL is authoritative on this worker.
 //    (another worker changed the intent, but this worker is responsible for the ACL, so update it.)
 // 2) ACL change - Intent may be anything - ACL just became authoritative on this worker.
 //    (this worker just became responsible, so check to make sure intent and ACL agree.)
 // 3) AuthorityIntent change - Intent is authoritative on this worker but no longer assigned to this worker - ACL is authoritative on this worker.
 //    (this worker had responsibility for both and is giving up authority.)
-void SpatialLoadBalanceEnforcer::QueueAclAssignmentRequest(const Worker_EntityId EntityId)
+// Queuing an ACL assignment request may not occur if the assignment is the same as before, or if the request is already queued.
+void SpatialLoadBalanceEnforcer::MaybeQueueAclAssignmentRequest(const Worker_EntityId EntityId)
 {
-	// TODO(zoning): measure the performance impact of this.
-	if (AclWriteAuthAssignmentRequests.ContainsByPredicate([EntityId](const WriteAuthAssignmentRequest& Request) { return Request.EntityId == EntityId; }))
+	const SpatialGDK::AuthorityIntent* AuthorityIntentComponent = StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(EntityId);
+	if (AuthorityIntentComponent == nullptr)
 	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("An ACL assignment request already exists for entity %lld on worker %s."), EntityId, *WorkerId);
+		// We should have checked the existence of the auth intent component before calling this function.
+		checkNoEntry();
+		return;
 	}
-	else
+
+	const PhysicalWorkerName* OwningWorkerId = VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
+	if (OwningWorkerId != nullptr &&
+		*OwningWorkerId == WorkerId &&
+		StaticComponentView->HasAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
+	{
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("No need to queue newly authoritative entity %lld because this worker is already authoritative."), EntityId);
+		return;
+	}
+
```

Comment:
Do we need an additional check that OwningWorkerId is actually valid?

Reply:
from a slack conversation with @alastairdglennie I believe this is now guaranteed to be fine. Please correct me if I'm wrong

Reply:
I have no memory of this conversation, let me TAL

Reply:
I agree that we don't need the OwnerWorkerId set. This function should only be called once we're past processing startup ops, at which point we know the Translator is ready, meaning our local worker ID exists in the Translator. Could maybe put a `check()` after the OwningWorkerId assignment for `OwningWorkerId != nullptr` rather than in the conditional?

Reply:
we now have a check, and an error + return if it happens in a non-dev build. This also means getting rid of the tests that set up the enforcer in such a way that it fails.

Reply:
Agreed, this was necessary before we added in the delay for the Translator to become ready, it should be safe now. The only remaining concern I would have would be for entities without an authority intent, but I think we've prevented that now too.